Add an API for Syncing Drafts Across the Zulip Server

Hypro999 opened the pull request:
One this is complete, Zulip clients will be able to create, update, fetch, and delete drafts from the backend allowing drafts to be synced across clients in the long run.

Related discussion on CZO: https://chat.zulip.org/#narrow/stream/2-general/topic/drafts
Related Issue on GitHub: https://github.com/zulip/zulip/issues/3775

Hypro999 commented:
Currently, the work done includes creating the model class and an endpoint for creating draft(s). The endpoint has unit test but the documentation for it is pending (I plan on doing all of the documentation work at once at the end).

Pending:
[ ] Add an endpoint for updating drafts
[ ] Add an endpoint for deleting drafts
[ ] Add an endpoint for fetching drafts
[ ] Documenting all of the endpoints

We might need some further discussion on what kinds of ways the user can fetch drafts. For example, either by type, or by topic, or by N most recent, or so on.

After this PR gets merged, we should have a round of discussion about how we will want to upgrade our clients for this. @Gittenburg I'd really like some help deciding how the UI would change to accommodate this when the time comes.

Hypro999 commented:
(force pushed to fix linting errors)

timabbott commented on zerver/lib/timestamp.py:
```diff
@@ -40,3 +41,97 @@ def timestamp_to_datetime(timestamp: float) -> datetime.datetime:
+        timestamp = dt.isoformat()
+        if timestamp.endswith("+00:00"):
+            timestamp = timestamp[:-6] + "Z"
+        return timestamp
```
Is this something we can't just do via pytz?  I don't really want to be maintaining a giant block of timezone-conversion code if we don't need to.

(Also, we get to pick what date format to use in Zulip APIs, right?  Seems like it might be better to just use UNIX timestamps if so like we do in other places)

Hypro999 replied:
Sure then, let's use Unix timestamps here to keep things consistent.

Hypro999 commented:
**Changes:** Changed `last_edit_time` to use Unix timestamps, and removed the original first commit which added extra code for RFC 3339.

Hypro999 commented:
(Bionic test failure is a flake)

Hypro999 commented:
**Changes:** I've added all of the CRUD API endpoints. I've also changed the commit for `create_drafts` a little because I needed to extract the validation logic. I've also renamed the "last_edited_date" API parameter to "timestamp" (I also return drafts to the user with this same key name).

All that's left is documenting these endpoints.

@timabbott can you review this now?

In particular, I had 3 questions:
1. Is `further_validated_draft_dict` a decent name or does someone have any better ideas? I feel like it could be better.
2. Is there any other kind of sanitization that would be prefered?
3. Do we want to hook this up to the events system? We might need more discussion on CZO about the "optionality" of this feature. @showell I'd like your opinion on this :slightly_smiling_face:

Hypro999 commented:
**changes:** Just fixed the capitalization in a few strings for error messages.

Hypro999 commented:
The test failure seems irrelevant. @hashirsarwar this seems to be related to something you were working on recently, mind taking a look?
```
Jul 24 03:56:18 ======================================================================
Jul 24 03:56:18 FAIL: test_push_bouncer_api (zerver.tests.test_push_notifications.PushBouncerNotificationTest)
Jul 24 03:56:18 This is a variant of the below test_push_api, but using the full
Jul 24 03:56:18 ----------------------------------------------------------------------
Jul 24 03:56:18 Traceback (most recent call last):
Jul 24 03:56:18   File "/usr/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
Jul 24 03:56:18     yield
Jul 24 03:56:18   File "/usr/lib/python3.8/unittest/case.py", line 676, in run
Jul 24 03:56:18     self._callTestMethod(testMethod)
Jul 24 03:56:18   File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
Jul 24 03:56:18     method()
Jul 24 03:56:18   File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/test/utils.py", line 373, in inner
Jul 24 03:56:18     return func(*args, **kwargs)
Jul 24 03:56:18   File "/usr/lib/python3.8/unittest/mock.py", line 1348, in patched
Jul 24 03:56:18     return func(*newargs, **newkeywargs)
Jul 24 03:56:18   File "/__w/zulip/zulip/zerver/tests/test_push_notifications.py", line 335, in test_push_bouncer_api
Jul 24 03:56:18     self.assertEqual(warn_log.output, [
Jul 24 03:56:18   File "/usr/lib/python3.8/unittest/case.py", line 912, in assertEqual
Jul 24 03:56:18     assertion_func(first, second, msg=msg)
Jul 24 03:56:18   File "/usr/lib/python3.8/unittest/case.py", line 1118, in assertListEqual
Jul 24 03:56:18     self.assertSequenceEqual(list1, list2, msg, seq_type=list)
Jul 24 03:56:18   File "/usr/lib/python3.8/unittest/case.py", line 1100, in assertSequenceEqual
Jul 24 03:56:18     self.fail(msg)
Jul 24 03:56:18   File "/usr/lib/python3.8/unittest/case.py", line 753, in fail
Jul 24 03:56:18     raise self.failureException(msg)
Jul 24 03:56:18 AssertionError: Lists differ: ['ERROR:django.request:Bad Gateway: /json/u[22 chars]ken'] != ['WARNING:root:Received 500 from push notif[82 chars]ken']
Jul 24 03:56:18 
Jul 24 03:56:18 First differing element 0:
Jul 24 03:56:18 'ERROR:django.request:Bad Gateway: /json/users/me/apns_device_token'
Jul 24 03:56:18 'WARNING:root:Received 500 from push notification bouncer'
Jul 24 03:56:18 
Jul 24 03:56:18 Second list contains 1 additional elements.
Jul 24 03:56:18 First extra element 1:
Jul 24 03:56:18 'ERROR:django.request:Bad Gateway: /json/users/me/apns_device_token'
Jul 24 03:56:18 
Jul 24 03:56:18 + ['WARNING:root:Received 500 from push notification bouncer',
Jul 24 03:56:18 - ['ERROR:django.request:Bad Gateway: /json/users/me/apns_device_token']
Jul 24 03:56:18 ? ^
Jul 24 03:56:18 
Jul 24 03:56:18 +  'ERROR:django.request:Bad Gateway: /json/users/me/apns_device_token']
Jul 24 03:56:18 ? ^
Jul 24 03:56:18 
Jul 24 03:56:18 
Jul 24 03:56:18 ----------------------------------------------------------------------
Jul 24 03:56:18 Ran 1546 tests in 106.501s
Jul 24 03:56:18 
Jul 24 03:56:18 FAILED (failures=1, skipped=28)
```

Hypro999 commented:
**changes:** Finished adding the documentation for all endpoints. This is no longer a draft but ready for a serious round of review.

timabbott commented:
This is nice work @Hypro999!  I merged all but the last commit, which has a few copy-pasted strings (will note in comments).  

Before working too much on the frontend, we'll also want to add `/register` / `page_params` support for including drafts, so that one doesn't need an extra round-trip to get them.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -448,6 +448,206 @@ paths:
+      description: |
+        Fetch all drafts for the current user.
+
+        `GET {{ api_url }}/v1/messages`
```
Endpoint URL is wrong :)

Hypro999 replied:
How embarrassing :man_facepalming:. I should have double checked this, looks like I got a little lazy. Thanks for the careful review here!

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -448,6 +448,206 @@ paths:
+                        description: |
+                          The number of drafts the user currently has. Also the
+                          number of drafts returned under "drafts".
+                        example: 3
```
Hmm, this `count` field feels unnecessary; the client can easily compute that.  Maybe better to just remove it?

Hypro999 replied:
We could, and I agree that we don't 100% need it. But what if the client (not necessarily our web/mobile clients) didn't want to have to loop over all of the elements before figuring out how many drafts were returned? Is there any harm in keeping this?

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -448,6 +448,206 @@ paths:
+        - name: drafts
+          in: query
+          description: |
+            A json-encoded list of all of the drafts that should be created for the current user.
```
JSON is usually capitalized in strings.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -448,6 +448,206 @@ paths:
+      operationId: create_drafts
+      tags: ["drafts"]
+      description: |
+        Create one of more drafts for the current user.
```
"one or more"

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -448,6 +448,206 @@ paths:
+                        type: array
+                        description: |
+                          An array of the IDS for the drafts that were just created in the same
+                          order as they were requested.
```
Probably we should document here that if any draft has a validation error, the entire request will fail.

Hypro999 replied:
Good idea.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -448,6 +448,206 @@ paths:
+      operationId: edit_draft
+      tags: ["drafts"]
+      description: |
+        Edit a single draft.
```
We should probably link to https://zulipchat.com/help/view-and-edit-your-message-drafts here.

Hypro999 replied:
Before we do that I want us to discuss how the clients will actually use this (more information below).

andersk commented on zerver/lib/timestamp.py:
```diff
@@ -40,3 +41,7 @@ def timestamp_to_datetime(timestamp: float) -> datetime.datetime:
     verify_UTC(dt)
     return calendar.timegm(dt.timetuple())
+
+def datetime_to_precise_timestamp(dt: datetime.datetime) -> float:
+    verify_UTC(dt)
+    return time.mktime(dt.timetuple()) + dt.microsecond/1000000
```
Isn’t this just `dt.timestamp()`?

Hypro999 replied:
It is, thanks for pointing it out. Seems like I went about this in a roundabout way. I'll fix it.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -448,6 +448,206 @@ paths:
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Draft"
```
Should this schema be modified to not have an `id` field in it?

Hypro999 replied:
Sorry, I don't get what you're suggesting here. Could you please elaborate? The `draft_id` field is necessary in the schema since it's a parameter that's part of the path.

timabbott replied:
Ahh, OK, I had been concerned that this meant it needed to be in the dictionary as well.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -448,6 +448,206 @@ paths:
+                      result:
+                        type: string
+                        description: |
+                          Literally just the word "error".
```
I think it'd be better to use the standard JsonError schema rather than documenting what specific string we'll use.

andersk replied:
Also in general we should be using [`enum`](https://swagger.io/docs/specification/data-models/enums/) (as `JsonError` does) to document when a string has a specific value or a small set of specific values.

timabbott commented:
Pushed just the last commit back here; @orientor can you also give this a review?

Hypro999 commented:
@timabbott, before that review though, I have a question that I have that's kinda bugging me is "is it really ok to use JSON-encoded strings here instead of using actual JSON literals here (for `POST` and `PATCH`)?". I mean for most languages, this just means 1 extra line of code using an existing JSON library. But something about it just feels kinda ugly.

I originally wanted to use JSON objects/literals instead of strings everywhere but I don't think that plays well with Zulip's existing REQ system. AFAIK, I'd need to take the entire body as JSON into a single variable in the view function using `REQ(argument_type=body)` and then parse this - but I don't think this plays well with the existing OpenAPI testing/validation framework. REQ doesn't read the first level of keys and automatically extract parameters from it as view function arguments AFAIK.

Can I try to sort of "step over" the REQ system by taking the request body `REQ(argument_type=body)` and performing variable extraction and validation in the view functions themselves?

Or is there already a well known endpoint that already does what I'm trying to do in a more consistent way?

andersk commented on zerver/lib/timestamp.py:
```diff
@@ -44,4 +43,4 @@ def datetime_to_timestamp(dt: datetime.datetime) -> int:
 
 def datetime_to_precise_timestamp(dt: datetime.datetime) -> float:
     verify_UTC(dt)
-    return time.mktime(dt.timetuple()) + dt.microsecond/1000000
+    return dt.timestamp()
```
The `verify_UTC` is unnecessary; `.timestamp()` works fine with timezones. And then this is not worth having as a helper function—the caller should just use `.timestamp()` directly.

timabbott replied:
I'd also wonder whether we actually want microsecond precision here when we don't use it elsewhere.

timabbott commented:
> Can I try to sort of "step over" the REQ system by taking the request body REQ(argument_type=body) and performing variable extraction and validation in the view functions themselves?

> Or is there already a well known endpoint that already does what I'm trying to do in a more consistent way?

The short answer is you shouldn't do that for now.  I think there's a world where we extend REQ to support a single JSON-encoded blob, rather than several parameters (either in addition to or in an `/api/v2` namespace replacing the existing system).  There's a good chance we'll decide to do that down the line because it's cleaner to marshall data with that sort of API.  But it certainly doesn't make sense to hack that for a single endpoint; much better would be to extend `zerver/lib/request.py` to support that system.

Could be a fine next project.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -1293,6 +1293,190 @@ paths:
+      description: |
+        Fetch all drafts for the current user.
+
+        `GET {{ api_url }}/v1/drafts`
```
Can you add `**Changes** and API_FEATURE_LEVEL for these new endpoints?

Hypro999 replied:
Would this be considered a 4.0 feature? I'm not sure about where to exactly put these changes. Also would the `API_FEATURE_LEVEL` become 31 or something else?

timabbott replied:
Yes, exactly, all new features go under the 4.0 heading until we release 4.0 -- and the API_FEATURE_LEVEL just gets incremented by 1 for each feature.

Hypro999 commented:
Oops, looks like `3b44ad4` was a mistake (I must have accidentally modified a previous commit). Let get rid of that.

Hypro999 commented:
So here are the changes that I just pushed:
1. Removed the first "precise timestamp" commit since we can just use `.timestamp()` directly. @andersk thanks for pointing this out! 😁
2. I modified the commit message of the documentation commit a bit (the "small related sidenote" stuff).
3. I fixed the OpenAPI documentation for the `POST /drafts`:
```diff
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 553fa1e2d5..032a20654f 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -1597,10 +1597,12 @@ paths:
           in: query
           description: |
             A JSON-encoded list of all of the drafts that should be created for the current user.
-          schema:
-            type: array
-            items:
-              $ref: "#/components/schemas/Draft"
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: "#/components/schemas/Draft"
       responses:
         "200":
           description: Success.
```

@orientor and @timabbott  can you guys review this commit once you get some free time later?

Hypro999 commented:
**changes:** Rebased the branch but otherwise left the documentation commit untouched. Then I added a few new commits based on recent discussions on CZO about how to let users decide to opt in or out. Ultimately what was decided was that drafts syncing is going to be a user-level feature (and not a device or draft-per-draft level feature).

Hypro999 commented:
Here are some of the things left before we can call the backend for this thing complete:
- Auto clear‌ing of drafts after a given period of time.
- Handling attachments.
- Event system integration - for both `drafts_syncing_enabled` as well as when something happens to a User's draft(s).
- Bump `API_LEVEL` and update the changelog.
- Create an endpoint to let the user update `drafts_syncing_enabled` . When set to False, delete all existing drafts.
- Make the number of recent drafts returned something that can be configured as a setting.

orientor commented on zerver/openapi/zulip.yaml:
```diff
@@ -1522,6 +1522,192 @@ paths:
+              schema:
+                $ref: "#/components/schemas/Draft"
+          description: |
+            A JSON-encoded string denoting the edited data you
```
I think  "JSON-encoded object" will provide more clarity.

Hypro999 commented:
**changes:** Pushed the documentation change suggested by @orientor (change "JSON-encoded string" to "JSON-encoded object").

orientor commented:
@Hypro999 other than a minor change the PR LGTM. One thing I am not sure about is that I have documented [/register](https://github.com/zulip/zulip/pull/16000) in a PR so I think draft properties for register will have to be documented too. @timabbott Thoughts? One of the ways can be that this PR gets merged before and then I can add a commit to the register PR for the three draft properties in /register.

Hypro999 commented:
**changes:** Made timestamp an integer instead of a float. ref: https://github.com/zulip/zulip/pull/16003#issuecomment-668333296

Hypro999 commented:
**changes:** Fixed the new commit (`Add endpoints for...`), then rebased this PR against master to resolve conflict in `zerver/openapi/zulip.yaml`.

Hypro999 commented:
**changes:** Following the advice from around [here](https://chat.zulip.org/#narrow/stream/2-general/topic/drafts/near/973662), `draft_dict`s will now contain the ID of the draft. Then since they now do this, we change the structure of the `drafts` parameter in the `GET /drafts` endpoint to be an array of drafts instead of an object mapping IDs to drafts.

Hypro999 commented:
@timabbott this PR is close to done, can you please give it follow-up round of review? Once you're done, I'll add one more commit bumping the API_LEVEL and updating the changelog. Then we can add [follow ups](https://github.com/zulip/zulip/pull/15880#issuecomment-667709844) in later PRs.

Hypro999 commented:
Ok, all rebased and ready for review.

Hypro999 commented:
Just rebased this PR against master, nothing new was added since the last comment.

Hypro999 commented:
**changes:** Instead of having two endpoints for enabling and disabling syncing drafts respectively (which in all honesty was a disgusting way of doing things), we now have a single endpoint for changing any kind of draft setting whatsoever (similar in style to the change notification settings endpoint). Honestly, I really just should have done this since the beginning.

Hypro999 commented:
I made a separate branch and carefully cherry-picked each commit by hand to produce a clean and modular commit history and polish anything I wanted/needed to along the way. Then I reset this branch to that clean one. I'm glad I took the time to do this since now I actually feel *fully satisfied with how clean this is.

Hypro999 commented:
I just need to add a bit more documentation:
1. Update the user-facing API docs.
2. Update the changelog.

Hypro999 commented:
Re-based this for the time being. I'll add the finishing touches of documentation in the morning (i.e. that changelog and public API documentation stuff).

Hypro999 commented:
Re-numbered migrations.

Hypro999 commented:
Nothing new with this push. Just rebased and renumbered migrations.

Hypro999 commented:
Pushed a small fix removing a sibling of $ref that was going to be ignored.

Hypro999 commented:
@timabbott can you review this? I fear that it's starting to go stale...

Hypro999 commented:
Again?

timabbott commented on zerver/models.py:
```diff
@@ -1025,6 +1025,9 @@ class UserProfile(AbstractBaseUser, PermissionsMixin):
     EMAIL_REMINDER_WAITPERIOD = 24
 
+    # Whether or not the user wants to sync their drafts.
+    drafts_syncing_enabled = models.BooleanField(default=False)
```
I think we should call this `enable_drafts_synchronization`, and default it to True; I imagine it'll be rare that folks turn this off, and we'll save a migration in the future.

timabbott commented on zilencer/management/commands/populate_db.py:
```diff
@@ -326,6 +326,14 @@ def handle(self, **options: Any) -> None:
+
+            zoe = get_user_by_delivery_email("ZOE@zulip.com", zulip_realm)
+            zoe.drafts_syncing_enabled = True
+            zoe.save(update_fields=["drafts_syncing_enabled"])
```
I think we can skip this logic and have the feature enabled for all default users.

timabbott commented on zerver/views/drafts.py:
```diff
@@ -85,11 +87,22 @@ def further_validated_draft_dict(draft_dict: Dict[str, Any],
+        if not user_profile.drafts_syncing_enabled:
+            return json_error(_("User has not enabled drafts syncing."))
+        return view_func(request, user_profile, *args, **kwargs)
+    return cast(ViewFuncT, draft_view_func)  # https://github.com/python/mypy/issues/1927
```
Hmm, what's your thinking on having this be a server-side access check, vs. a client-side decision on whether to try to send drafts to the server?  I'd generally expect this sort of check to be client-side, but I could could see an argument to go the other way.  (Do we want to have the `do_...` function in `actions.py` remove all existing drafts when one disables the synchronization setting?  Probably)

In any case, we probably want the `human_users_only` decorator on these views as well (to disable bot access).

Hypro999 replied:
@timabbott  good question! There will definitely be client side checks for this but I've added a server side check as well as a sort of defensive measure.

[Feedback](https://chat.zulip.org/#narrow/stream/2-general/topic/drafts/near/931751) from @showell made me realize that if someone doesn't want to sync their drafts, they can explicitly tell the server they don't want drafts to be synced. This way even if a client is kinda accidentally buggy then drafts won't be synced (though this isn't the case if it straight up ignores what the user wants since it can always turn this flag back on).

Hypro999 replied:
> Do we want to have the do_... function in actions.py remove all existing drafts when one disables the synchronization setting? Probably

Yup, and that's exactly what we're doing. Ref: `do_disable_drafts_syncing` from `zerver.lib.drafts`

Hypro999 replied:
> In any case, we probably want the human_users_only decorator on these views as well (to disable bot access).

I've actually thought about this before and I think it might be better to allow bots to have their own drafts. This will allow one to get more hackish with what they can do with a bot (which is actually a good thing).

Think about it this way: drafts can be used like templates since the message sending and drafts deletion logic are not really coupled (you could send a message but you don't have to delete the draft - it all depends on the client). So theoretically, one could use drafts as a way of creating message templates for bots which will be stored on the server instead of locally. Following this line of thought, I felt that it would be better to let bots access this endpoint too.

Just to be clear: I don't have a **super** concrete use case for allowing bots to use drafts but rather I have a general feeling that it would be better.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -7930,6 +7930,50 @@ paths:
                         "msg": "",
                         "result": "success",
                       }
+  /settings/drafts:
```
I think rather than creating additional `/settings/` endpoints, we should probably just add this feature to the main `/settings` endpoint.  (the same argument holds for the existing `enter_sends`, which if you're up for it would be well worth migrating in a separate prep PR to figure out how to extend that endpoint without the complications of a new feature).  Overall it was probably a mistake to introduce separate /settings/notifications and /settings/display endpoints, but at least those are 10+ related settings each; creating a new endpoint with just 1 setting seems like it's going to clutter our API documentation for little to no benefit.

The other approach I could see taking is making a `/settings/compose` to be a bit broader than just drafts, but I'd be hard-pressed to come up with 10 more settings in that category.

timabbott commented:
Sorry for the delayed review on this @Hypro999!  I merged the first 3 commits (through 99cf37dc511ae6458e2f7fb8e960260d39e89c8b) and posted a batch of comments that should let us proceed with merging the next few.  

Strategically I think it's worth doing the one-time migration to setup `POST /settings` as an endpoint supporting the `property_types` framework as a container for the "drafts enabled" migration as discussed in https://github.com/zulip/zulip/pull/15880#pullrequestreview-519952454; I think we should do that as a separate PR just to avoid cluttering this drafts PR with the back-and-forth about the precise design of that.  Sound good?

Hypro999 commented:
> I think we should do that as a separate PR just to avoid cluttering this drafts PR with the back-and-forth about the precise design of that. Sound good?

So are you saying that we'll go with the way things are for now (with respect to the `drafts_enabled` setting) then handle migrating all of the `/settings/` endpoints to `/settings` later in a separate PR? If so then I definitely agree. Otherwise could you please clarify what you meant?

Hypro999 commented:
(Just rebased with master)

Hypro999 commented:
Rebased against master because of the recent code-formatting changes (migration to psf/black).

Hypro999 commented:
Whoops, looks like I messed up a rebase and ended up pushing master back to this branch instead...

Hypro999 commented:
For anyone still interested in this, I've recovered the code and it's available at: https://github.com/hypro999/zulip/tree/issue-3375.

I'll open a new PR if and when it's needed.
